google/cloud/internal/service_endpoint.cc

```diff
@@ -0,0 +1,61 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/internal/service_endpoint.h"
+#include "google/cloud/internal/absl_str_cat_quiet.h"
+#include "google/cloud/internal/make_status.h"
+#include "google/cloud/universe_domain_options.h"
+#include "absl/strings/strip.h"
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
+
+// TODO(#13122): This function may need to be enhanced to support an env var
+// for UniverseDomain.
+StatusOr<std::string> DetermineServiceEndpoint(
+    absl::optional<std::string> endpoint_env_var,
+    absl::optional<std::string> endpoint_option, Options const& options,
+    std::string default_endpoint) {
+  auto constexpr kGoogleDefaultUniverse = ".googleapis.com.";
+  if (endpoint_env_var.has_value()) return *endpoint_env_var;
+  if (endpoint_option.has_value()) return *endpoint_option;
+  if (!absl::EndsWith(default_endpoint, ".")) {
+    absl::StrAppend(&default_endpoint, ".");
+  }
+  if (!absl::EndsWith(default_endpoint, kGoogleDefaultUniverse)) {
+    return internal::InternalError(
+        absl::StrCat("default_endpoint is not in GDU: ", default_endpoint));
```

Comment:
Nit:  Is it worth showing the user an acronym here, particular when (it seems) we're uncomfortable with `kGDU` in the code?